Added WPFGetting Started Guide (#661)
MicrosoftDocs/edge-developer: microsoft-edge/webview2/gettingstarted/wpf.md

---
description: Host web content in your WPF app with the Microsoft Edge WebView 2 control
title: Microsoft Edge WebView 2 for WPF apps
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 05/08/2020
ms.topic: reference
ms.prod: microsoft-edge
ms.technology: webview
keywords: IWebView2, IWebView2WebView, webview2, webview, wpf apps, wpf, edge, ICoreWebView2, ICoreWebView2Controller, browser control, edge html
---

# Getting Started with WebView2 in WPF (developer preview)

This walkthrough goes over the commonly used functionalities of [WebView2 (developer preview)](https://aka.ms/webview) and gets you started on creating your first WebView2 app. Visit [API reference](../reference/dotnet/0-9-494-reference-webview2.md) to learn more about individual APIs.  

## Prerequisites

* [Microsoft Edge (Chromium)](https://www.microsoftedgeinsider.com/download/) installed on supported OS (currently Windows 10, Windows 8.1, and Windows 7). **We recommend using the Canary channel and the minimum required version is 82.0.488.0**.
* [Visual Studio](https://visualstudio.microsoft.com/) 2015 or later with .NET support installed.

> 

Reviewer: Committing change prior to edit pass.

## Step 1 - Create a single window WPF application

> We will start with a basic desktop project containing a single main window.

Reviewer: Committing change prior to edit pass.

> 1. Open **Visual Studio 2019**

Reviewer: Committing change prior to edit pass.

2. Create a **WPF .NET Core App** or a **WPF .NET Framework App** by selecting the correct project type and select **Next** on the bottom right corner of the window.

    ![wpfcore](media/wpf-gettingstarted-wpfcore.png)

    ![wpffw](media/wpf-gettingstarted-wpffw.png)

> 

Reviewer: Committing change prior to edit pass.

> 3. Give your project a name and set a location to save the project to. For .NET Framework, select the version to support. **WebView2 currently supports .NET Framework 4.6.2+ and .NET Core 3.0+**. Press **Create** in the bottom right corner when you're done.

Reviewer: Committing change prior to edit pass.

    ![createcore](media/wpf-gettingstarted-createcore.png)

    ![createfw](media/wpf-gettingstarted-createfw.png)

## Step 2 - Install WebView2 SDK

> Now let's add the WebView2 SDK into the project. For the developer preview, you can install the Win32 SDK via Nuget.

Reviewer: Committing change prior to edit pass.

> 1. Right click the project and click **Manage Nuget Packages**.

Reviewer: Committing change prior to edit pass.

    ![nuget](media/wpf-gettingstarted-mngnuget.png)

> 2. Enter **Microsoft.Web.WebView2** in the search bar, click **Microsoft.Web.WebView2** from the results, set the package version to **pre-release**, and click **Install** in the right hand side window to install the latest SDK. Nuget will download the SDK to your machine.

Reviewer: Committing change prior to edit pass.

>     ![manageNugetPackages](../media/manage-nuget-packages.png)

Reviewer: Added not inline, closing suggestion.

> You are all set to use and build against the WebView2 API. Press F5 to build and run the sample app. You should see an app displaying an empty window.

Reviewer: Committing change prior to edit pass.

![emptyApp](media/wpf-gettingstarted-blank.png)

## Step 3 - Create a single WebView in MainWindow.xaml

> Now let's add a WebView to the main window.

Reviewer: Committing change prior to edit pass.

1. Open the **MainWindow.xaml** file. Add the **WebView2** XML namespace by inserting the following line inside the `<Window/>` tag:

```h
xmlns:wv2="clr-namespace:Microsoft.Web.WebView2.Wpf;assembly=Microsoft.Web.WebView2.Wpf"
```

> **MainWindow.xaml** should look like the following:

Reviewer: Committing change prior to edit pass.

```xml
<Window x:Class="WPF_Getting_Started.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:{YOUR PROJECT NAME}"

        xmlns:wv2="clr-namespace:Microsoft.Web.WebView2.Wpf;assembly=Microsoft.Web.WebView2.Wpf"

        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800">
    <Grid>

    </Grid>
</Window>
```

> 3. Add the WebView2 control by replacing the `<Grid></Grid>` tags, with the following code snippet. The **Source** property sets the initial URI the WebView navigates to.

Reviewer: Committing change prior to edit pass.

```xml  
<DockPanel>
    <wv2:WebView2
        Name="webView"
        Source="https://www.microsoft.com"
    />
</DockPanel>
```

Press F5 to build and run the app. Now you have a WebView window displaying <https://www.microsoft.com>.

![microsoft](media/wpf-gettingstarted-microsoft.png)

## Step 4 - Navigation

Let's add the ability to navigate the WebView to the URL displayed in the search bar.

> 

Reviewer: Committing change prior to edit pass.

1. In **MainWindow.xaml**, add a Search Bar by copying and pasting the following code snippet inside the DockPanel that holds the WebView.

```xml
<DockPanel DockPanel.Dock="Top">
    <Button x:Name="ButtonGo" DockPanel.Dock="Right" Click="ButtonGo_Click" Content="Go"/>
    <TextBox Name="searchBar"/>
</DockPanel>
```

> It should look like this:

Reviewer: Committing change prior to edit pass.

```xml
    <DockPanel>
        <DockPanel DockPanel.Dock="Top">
            <Button x:Name="ButtonGo" DockPanel.Dock="Right" Click="ButtonGo_Click" Content="Go"/>
            <TextBox Name="searchBar"/>
        </DockPanel>
        <wv2:WebView2
            Name = "webView"
            Source = "https://www.microsoft.com"
        />
    </DockPanel>
```

> 2. Open **MainWindow.xaml.cs** in Visual Studio. Add the **CoreWebView2** namespace by copying and pasting the following statement at the top of **MainWindow.xaml.cs**.

Reviewer: Committing change prior to edit pass.

```csharp
using Microsoft.Web.WebView2.Core;
```

> 3. In **MainWindow.xaml.cs**, define the `ButtonGo_Click` method to navigate the WebView to the URI in the search bar.

Reviewer: Committing change prior to edit pass.

## Step 6 - Scripting

The hosting app can also inject JavaScript into WebView. You can task WebView to execute arbitrary JavaScript or add initialization scripts. Added initialization scripts apply to all future top level document and child frame navigation until removed, and run after the global object has been created and before any other script included by the HTML document is executed.

We can use Scripting to alert the user if they are navigating to a non-https site. Modify the `EnsureHTTPS` function it injects script into the web content using the [ExecuteScriptAsync](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#executescriptasync) method.

```csharp
void EnsureHTTPS(object sender, CoreWebView2NavigationStartingEventArgs args)
{
    String uri = args.Uri;
    if (!uri.StartsWith("https://"))
    {
        webView.CoreWebView2.ExecuteScriptAsync($"alert('{uri} is not safe, try an https link')");
        args.Cancel = true;
    }
}
```

Press F5 to build and run the app. Now if you try to navigate to a non-https site, the application alerts you that doing so is not possible.

![https](media/wpf-gettingstarted-https.png)

> ## Step 7 - Communication between host and web content

Author: there is only one host

> The host and the web content can also communicate with each other through `postMessage`. The web content running within a WebView can post to the host through `window.chrome.webview.postMessage`, and the message would be handled by any registered `ICoreWebView2WebMessageReceivedEventHandler` on the host. Likewise, the host can message the web content through `CoreWebView2.PostWebMessageAsString` or `CoreWebView2.PostWebMessageAsJSON`, which would be caught by handlers added from `window.chrome.webview.addEventListener`. The communication mechanism allows the web content to utilize native capabilities by passing messages to ask the host to call native APIs.

Reviewer 2: ```suggestion
Hosts and web content may communicate with each other using `postMessage` as follows:

- Web content in a WebView2 control may post a message to the host using `window.chrome.webview.postMessage`. The host handles the message using any registered `ICoreWebView2WebMessageReceivedEventHandler` on the host. - Hosts can posts messages to web content in a WebView2 control using `CoreWebView2.PostWebMessageAsString` or `CoreWebView2.PostWebMessageAsJSON`. These messages are caught by handlers added to `window.chrome.webview.addEventListener`. 

This communication mechanism allows web content to pass messages to the host using native capabilities.

```

> As an example to understand the mechanism, when the WebView2 navigates to a URI, lets display that URI in the search bar and alert the user of the URI displayed in the WebView

Reviewer 2: ```suggestion
In your project, when the WebView2 control navigates to a URL, you'll display the URL in the search bar and alert the user of the URL displayed in the WebView2 control.
```

> 1. Web Messaging is a **CoreWebView2** functionality. Because the initialization of **CoreWebView2** is asynchronous, create an `InitializeAsync` function that awaits [EnsureCoreWebView2Async](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#ensurecorewebview2async). In **MainWindow.xaml.cs** update your constructor and define an `InitializeAsync` function:

Reviewer 2: ```suggestion
1. In **MainWindow.xaml.cs**, update your constructor and create an `InitializeAsync` function as shown in the following code snippet. The `InitializeAsync` function awaits [EnsureCoreWebView2Async](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#ensurecorewebview2async) because the initialization of `CoreWebView2` is asynchronous.
```

```csharp
public MainWindow()
{
    InitializeComponent();
    webView.NavigationStarting += EnsureHTTPS;
    InitializeAsync();
}

async void InitializeAsync()
{
    await webView.EnsureCoreWebView2Async(null);
}
```

> 2. After **CoreWebView2** is initialized, register an event handler to respond to **Web Message Received**. In this example, the host does three things:
>     1. Receives a web message (the URI) from the web content using [TryGetWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2webmessagereceivedeventargs#trygetwebmessageasstring).
>     2. Updates the text in the search bar to the URI
>     3. Returns the message (the URI) back to the web content using [PostWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2#postwebmessageasstring)
> 
> In **MainWindow.xaml.cs** update `InitializeAsync` and define `UpdateSearchBar` as follows:
> 
> ```csharp
> async void InitializeAsync()
> {
>     await webView.EnsureCoreWebView2Async(null);
>     webView.CoreWebView2.WebMessageReceived += UpdateSearchBar;
> }
> 
> void UpdateSearchBar(object sender, CoreWebView2WebMessageReceivedEventArgs args)
> {
>     String Uri = args.TryGetWebMessageAsString();
>     searchBar.Text = Uri;
>     webView.CoreWebView2.PostWebMessageAsString(Uri);
> }
> 
> ```

Reviewer 2: ```suggestion

2. After **CoreWebView2** is initialized, register an event handler to respond to `WebMessageReceived`. In **MainWindow.xaml.cs** update `InitializeAsync` and add `UpdateAddressBar` using the following code snippet. 

```csharp
async void InitializeAsync()
{
    await webView.EnsureCoreWebView2Async(null);
    webView.CoreWebView2.WebMessageReceived += UpdateAddressBar;
}

void UpdateAddressBar(object sender, CoreWebView2WebMessageReceivedEventArgs args)
{
    String Uri = args.TryGetWebMessageAsString();
    addressBar.Text = Uri;
    webView.CoreWebView2.PostWebMessageAsString(Uri);
}

```
In the above code snippet, the host performs the following actions:
1. Receives a web message (the URL) from the web content using [TryGetWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2webmessagereceivedeventargs#trygetwebmessageasstring).
2. Updates the text in the search bar to the URL.
3. Returns the message (the URL) to the web content using [PostWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2#postwebmessageasstring)

```

3. In order for the WebView to send and respond to the web message, after **CoreWebView2** is initialized, the host:

    1. Injects a script to the web content that registers a handler to print message from the host
    2. Injects a script to the web content that posts the URL to the host

> 

Reviewer 2: Can you review and clarify these steps please?

Reviewer 2: The code and the numbered list are not in sync. Can you switch #1 and #2 around in the numbered list? Also can you change "injects a script" to "adds code" ?

Author: injects script is the write technical terminology for what is going on here - the numbering has already been removed

In **MainWindow.xaml.cs** update `InitializeAsync` as follows:

```csharp
async void InitializeAsync()
{
    await webView.EnsureCoreWebView2Async(null);

>     webView.CoreWebView2.WebMessageReceived += UpdateSearchBar;

Reviewer 2: ```suggestion
    webView.CoreWebView2.WebMessageReceived += UpdateAddressBar;
```

    await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.chrome.webview.postMessage(window.document.URL);");
    await webView.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync("window.chrome.webview.addEventListener(\'message\', event => alert(event.data));");
}
```

Press F5 to build and run the app. Now the search bar displays the URI in the WebView and when you successfully navigate to a new URI, the WebView alerts the user of the URI displayed in the WebView.

> ![searchbar](media/wpf-gettingstarted-searchbar.png)

Reviewer 2: Can you review the text and image - is a confirmation step missing?

Congratulations, you've just built your first WebView2 app!

## Next Steps

There are plenty of WebView2 functionalities that are not covered in this walkthrough.

To learn more:

* Please explore [API reference](../reference/dotnet/0-9-494-reference-webview2.md) for detailed information about each API.  

> ## Feedback

Reviewer 2: ```suggestion
## Get in touch with the Microsoft Edge WebView team 
```

> Help us build a richer WebView2 experience by sharing your feedback! Visit our [feedback repo](https://aka.ms/webviewfeedback) to submit feature requests or bug reports or search for known issues.

Reviewer 2: Can you add a See Also section with 3-5 links to related content please?

Author: We have a Next Steps section. We will link to the sample apps and reference when they are published should be a simple change. They will be published tomorrow morning
